src/client/cli/util.js

```diff
@@ -44,8 +49,99 @@ function isB58Multihash (str: string): boolean {
   }
 }
 
+function setupSSHTunnel (config: Object): Promise<Object> {
+  return new Promise((resolve, reject) => {
+    sshTunnel(config, (err, server) => {
+      if (err) return reject(err)
+      resolve(server)
+    })
+  })
+}
+
+function sshTunnelFromDeployCredentialsFile (filePath: string): Promise<Object> {
+  const creds = loadDeployCredentials(filePath)
+  const config = deployCredsToTunnelConfig(creds)
+  return setupSSHTunnel(config)
+}
+
+function loadDeployCredentials (filePath: string): Object {
+  if (filePath.endsWith('.yaml')) {
+    return yaml.safeLoad(fs.readFileSync(filePath, 'utf8'))
+  } else if (filePath.endsWith('.json')) {
+    return JSON.parse(fs.readFileSync(filePath, 'utf8'))
+  } else {
+    throw new Error('Loading deploy credentials is only supported from .json or .yaml files')
+  }
+}
+
+function deployCredsToTunnelConfig (deployCreds: Object | string, extraConfigOptions: Object = {}): Object {
+  if (typeof deployCreds === 'string') {
+    deployCreds = loadDeployCredentials(deployCreds)
+  }
+  const {ip: host, vps_user: {name: username, password}} = deployCreds
+  if (host == null || username == null || password == null) {
+    throw new Error('Deploy credentials are in unexpected format')
+  }
+  const opts = {
+    username,
+    password,
+    host,
+    dstPort: '9002',
+    localHost: 'localhost',
+    keepAlive: true
+  }
+  return Object.assign({}, opts, extraConfigOptions)
+}
+
+type GlobalOptions = {
+  apiUrl: string,
+  deployCredentialsFile?: Object
+}
+
+type SubcommandGlobalOptions = { // eslint-disable-line no-unused-vars
+  client: RestClient
+}
+
+function subcommand<T: SubcommandGlobalOptions> (handler: (argv: T) => Promise<*>): (argv: GlobalOptions) => void {
+  return (argv: GlobalOptions) => {
+    const {apiUrl, deployCredentialsFile} = argv
+    const client = new RestClient({apiUrl})
+
+    const sshTunnelConfig = (deployCredentialsFile != null)
+      ? deployCredsToTunnelConfig(deployCredentialsFile)
+      : null
+
+    let sshTunnelPromise = Promise.resolve()
+    let sshTunnel = null
+    if (sshTunnelConfig != null) {
+      sshTunnelPromise = setupSSHTunnel(sshTunnelConfig)
+        .then(tunnel => { sshTunnel = tunnel })
+    }
+
+    function closeTunnel () {
+      if (sshTunnel != null) {
+        sshTunnel.close()
+      }
+    }
+
+    // Using lodash as a kind of flow "escape valve", since it's being stubborn
+    const subcommandOptions: T = set(clone(argv), 'client', client)
+
+    sshTunnelPromise
+      .then(() => handler(subcommandOptions))
+      .then(closeTunnel)
+      .catch(err => {
```

Comment:
Yeah, I wasn't happy about that either.  I think the right thing to do is have the command handlers just throw on error conditions and have the `subcommand` helper be responsible for printing errors to the console, cleaning up and exiting the process.